http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala

```diff
@@ -0,0 +1,367 @@
+package org.broadinstitute.dsde.workbench.leonardo.monitor
+
+import java.time.Instant
+
+import cats.effect.{Async, Sync, Timer}
+import cats.implicits._
+import cats.mtl.ApplicativeAsk
+import com.google.cloud.storage.BucketInfo
+import fs2.Stream
+import io.chrisdavenport.log4cats.Logger
+import org.broadinstitute.dsde.workbench.google2.{GcsBlobName, GoogleStorageService}
+import org.broadinstitute.dsde.workbench.leonardo._
+import org.broadinstitute.dsde.workbench.leonardo.dao.ToolDAO
+import org.broadinstitute.dsde.workbench.leonardo.db._
+import org.broadinstitute.dsde.workbench.leonardo.http.{ctxConversion, dbioToIO, nowInstant}
+import org.broadinstitute.dsde.workbench.leonardo.monitor.RuntimeMonitor.CheckResult
+import org.broadinstitute.dsde.workbench.leonardo.util.{RuntimeAlgebra, StopRuntimeParams}
+import org.broadinstitute.dsde.workbench.model.TraceId
+import org.broadinstitute.dsde.workbench.model.google.{GcsPath, GoogleProject}
+import monocle.macros.syntax.lens._
+import org.broadinstitute.dsde.workbench.leonardo.monitor.MonitorState._
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+abstract class BaseCloudServiceRuntimeMonitor[F[_]] {
+  implicit def F: Async[F]
+  implicit def timer: Timer[F]
+  implicit def dbRef: DbReference[F]
+  implicit def ec: ExecutionContext
+  implicit def runtimeToolToToolDao: RuntimeContainerServiceType => ToolDAO[F, RuntimeContainerServiceType]
+  implicit def openTelemetry: OpenTelemetryMetrics[F]
+
+  def runtimeAlg: RuntimeAlgebra[F]
+  def logger: Logger[F]
+  def googleStorage: GoogleStorageService[F]
+
+  def monitorConfig: MonitorConfig
+
+  def process(runtimeId: Long)(implicit ev: ApplicativeAsk[F, TraceId]): Stream[F, Unit] =
+    for {
+      // building up a stream that will terminate when gce runtime is ready
+      traceId <- Stream.eval(ev.ask)
+      startMonitoring <- Stream.eval(nowInstant[F])
+      monitorContext = MonitorContext(startMonitoring, runtimeId, traceId)
+      _ <- Stream.sleep(monitorConfig.initialDelay) ++ Stream.unfoldLoopEval[F, MonitorState, Unit](Initial)(s =>
+        Timer[F].sleep(monitorConfig.pollingInterval) >> handler(monitorContext, s)
+      )
+    } yield ()
+
+  def pollCheck(googleProject: GoogleProject,
+                runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig,
+                operation: com.google.cloud.compute.v1.Operation,
+                action: RuntimeStatus)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
+
+  private def handler(monitorContext: MonitorContext, monitorState: MonitorState): F[CheckResult] =
+    for {
+      now <- nowInstant
```

Comment:
Just curious if we need `MonitorContext` _and_ `ApplicationContext`.. could `MonitorContext` suffice?

Reply:
so `MonitorContext`'s now is fixed to when the monitoring starts, but `AppContext`'s now is more like when the function is being called, which might change throughout the monitoring process